### ISSUE TITLE
Review test page for verification of photo records

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#1379]
A test page was previously set up but not pursued at the time (this is to facilitate quick review and verification of records that have photos, minimising the need to click in and out of thumbnail images).

Reinstate the test page, carry out initial review and invite further testing from verifiers (initially through the Supporting Science project).

**Comment by johnvanbreda** [comment]
Page re-instated at https://irecord.org.uk/verification/photos - on live so it can't get overwritten again. 
Keyboard shortcuts should be same as the grid version. Things that might need work:
* There is currently no multi-select/bulk verify option.
* Consider which fields to show for each photo in the panel.

**Comment by kitenetter** [comment]
Test with users, need to define questions to ask

**Comment by kitenetter** [comment]
Here's some feedback from me, plus some possible options for further development - it would be good to know if these options are realistically achievable before we seek input from other verifiers as to whether they are needed.

1. At the moment iNaturalist records are not separated out, and I think they will need to be before we go live.
2. The map display takes up more space on the photo page than the normal page, which means that the record details pane is less visible, would prefer the map to be less deep.
3. Would very much like to be able to sort the records; in particular I would want to be able to sort by species name, but there may also be demand for sorting by other attributes e,g, recorder name
4. I think we should show the recorder name (as well as the date) in each photo panel
5. At the moment the "Edit" and "View" buttons for individual records do not function on the photo page.

I think we looked at this document some time ago:
[Photo verification mock-up MH 2021_05_21.pdf](https://github.com/BiologicalRecordsCentre/iRecord/files/10227258/Photo.verification.mock-up.MH.2021_05_21.pdf)

It contains a suggested workflow based on dealing with one image at a time, and having the verification buttons within the image panel, so that all a verifier would need to do is open up the first panel, click once to make a verification decision (and then save it), after which the next photo is automatically presented.

How much of what is suggested in that document is possible to develop?

**Comment by johnvanbreda** [comment]
I've done points 2, 4 and 5. I suggest we create the iNat page once the features are completely implemented and tested, since this is a simple filter and it saves updating 2 versions of the page. 

I still need to add the sort feature.

Features in the PDF file are now implemented in the develop branch of code.

**Comment by johnvanbreda** [comment]
Sort tool now implemented in develop code. Let me know if you are planning to test @kitenetter and I'll update everything - otherwise I'll continue with other changes and do the updates all in one go.

**Comment by kitenetter** [comment]
@johnvanbreda the plan is to invite a small number of verifiers to help test the page, so it probably makes sense to complete your current work on this first, unless there is anything that you think we should seek views on before committing to developing. For Supporting Science purposes it would be good to have something ready to test by mid January.

**Comment by johnvanbreda** [comment]
To summarise, the remaining tasks are:
* Release the code
* Add an iNat page version and filter iNat records from the main version.

**Comment by kitenetter** [comment]
@johnvanbreda some initial feedback:

**1 - token not working**
When trying to accept a record as "Correct", I used a template of:
`Thanks! That's a great record of {{ taxon full name }} which has been {{ action }}.`

This resulted in preview text:
`Thanks! That's a great record of Chrysopilus cristatus (Black Snipefly) which has been redetermined.`

I.e. the token for "{{ action }}" is not picking up on the relevant action at the point of use.

**2 - copying text** 
The use of ctrl + C to copy (from anywhere on the page including the verification templates) results in the current photo card closing or opening.

**3 - tabs**
When working with keyboard shortcuts on the main verification page I use this sequence:
- "1" to open the "Correct" window
- Tab to move to the Save button
- Enter to apply the Save button

However, the tab key now takes me to the new tabs for the templates. I think this just needs the tab sequence to be changed so that the first tab key press still goes to the Save button, and then subsequent tab key presses go to the template options.

**4a - template Help**
Suggest moving the example text so that it appears above the list of tokens (i.e. before "Available replacement tokens are as follows"):
`An example template might look like the following:`
`Thanks! That's a great record of {{ taxon full name }} which has been {{ action }}.`

**4b - template Help**
Suggest re-wording the example text to the more neutral:
`Thanks for your record of {{ taxon full name }} which has been {{ action }}.`

**5 - recorders not logged in**
There is little point is exposing the comment templates for users who are not logged in (since they will never see the comments). This particularly relates to iNat, so maybe something to think about when we separate out the iNat version of the page, but even within iRecord there are recorders who are not logged in. Possibly it would be worth disabling the template options for records whose recorder was not logged in?

**6 - iNat link missing**
Presumably this will be sorted when we split out a separate iNat version of the page.

**Comment by johnvanbreda** [comment]
@kitenetter I've managed to fix 1, 2, 4a, 4b.
Other points still to do.

**Comment by johnvanbreda** [comment]
Point 3 - tab ordering has been improved in the develop branch as the buttons are now positioned before the text area in the tab sequence instead of after.

**Comment by johnvanbreda** [comment]
5. I'm not sure I see the point of making the comment behaviour different if you are not logged in. The comment still gets attached to the record and may get seen at some point by someone and it just means the UI is inconsistent so may cause confusion. Also taking this thought to its logical conclusion, we may as well drop the comment entirely if the user is not logged in.
6. The iNat links seem OK to me - please can you confirm which record they were missing for?

**Comment by johnvanbreda** [comment]
Following on from user testing feedback, the test installation has had the following updates:
* Improved visibility of the border around each card, so 1 record with 2 images is clearly differentiated from 2 records with an image each.
* Double clicking a card maximises it.
* Single click on an unselected card image does not pop the image up, you need to select the card first.
* Improved layout, so the maximised card, panel container, map and details form will all fit on a single screen. Scrollbars used within the controls as needed.

**Comment by kitenetter** [comment]
@johnvanbreda on the live site, if you click on a verification button and then click "Cancel", the verification decision is being applied regardless. This is dangerous as some of our testers may be exploring the live site - can we correct the behaviour of the Cancel button?

**Comment by kitenetter** [comment]
The issue about the Cancel button applies to all verification pages so has been moved to #1423

**Comment by kitenetter** [comment]
Testing on the live site:

If I redet a record from an open card, the record disappears (as expected) and the entire page seems to get faded out, and it is difficult to see how to move on to the next card.

If I open a card and then use the keyboard shortcut "i" to open and close an image, the card is no longer 'live' and I can't go on to use the keyboard shortcut "1" to accept the record.

**Comment by kitenetter** [comment]
After working through a number of records, a 'white box' seems to have appeared in the bottom half of the grid, obscuring the lower part of the open card. I can scroll the card to see the details but it is not very convenient:

![image](https://user-images.githubusercontent.com/16039111/216722640-c5898e1c-3f63-4ac4-afd3-afb388ab93eb.png)

**Comment by kitenetter** [comment]
The links within each card for "view record details" and "view species details" are not updating as you move between cards, they always seem to link to the first card in the grid, even after that card has been verified and removed from the pending grid.

**Comment by johnvanbreda** [comment]
> [This happened once but I can't reproduce it: If I open a card and then use the keyboard shortcut "i" to open and close an image, the card is no longer 'live' and I can't go on to use the keyboard shortcut "1" to accept the record.}

- I've been unable to reproduce this.

**Comment by johnvanbreda** [comment]
All bugs reported above now fixed. Also completed the following which came out of user testing:
1. Popups are now modal dialogs that can be dragged to a new position.
2. Comments don't clear if you close the dialog and re-open, unless you moved to a new record or selected a different status.
3. Terminology in the help: token changed to placeholder.

Changes are on live and the test site. @kitenetter if you can check this is now functioning OK., then I can set up the iNat variation before we publish them.

**Comment by kitenetter** [comment]
Opened up the first card and worked through to the end of a set of records. The last record in the queue did not display as an open card, it only showed up as a closed card in the grid. When I accepted the last record from the grid, the verification box did not close. I think the record has been verified correctly but I am left with some doubt:

![image](https://user-images.githubusercontent.com/16039111/226140193-d46668ab-2cca-48e4-a71d-3641f12be731.png)

**Comment by kitenetter** [comment]
And following on from the previous item, I then clicked on "Cancel" in the "Set status" box, but from that point onwards no photos would display, even if I changed the Context filter to look at a different set of records. It was fine again after a page refresh but there appears to be some odd behaviour when the last record in a set is reached.

**Comment by kitenetter** [comment]
In fact, the last record in the set does not get flagged as Accepted, so it appears that the "Set status" box is not functioning for the last record in a set.

HAR file from repeating the process (the issue seems to be reproducible):

[irecord_2023_03_18.zip](https://github.com/BiologicalRecordsCentre/iRecord/files/11009627/irecord_2023_03_18.zip)

**Comment by kitenetter** [comment]
The photo verification page does not currently allow filtering by grid ref precision - at the moment we can only do this via the column-head filters, which aren't available on the photo page. We do need to be able to filter for grid ref precision in order to deal with iNat records.

**Comment by kitenetter** [comment]
We discussed whether it would be possible to allow bulk verification on the photo page, but I can't remember what decision we arrived at. I don't think it is possible at the moment. If it is an easy thing to add it would be good to do so, but if it isn't easy I think we should go ahead without it, and review later if there is demand for it.

**Comment by kitenetter** [comment]
When the iNat page is set up, the blue box that says "View details of this record in iNaturalist using the link above" needs to be edited. The link may be able above, or it may be in the photo card, so it's probably best to just say:
"View details of this record in iNaturalist using the link provided"

**Comment by johnvanbreda** [comment]
@kitenetter I've checked for conventions and it seems that an A-Z icon should have the arrow pointing up, so I've corrected the sort button (e.g. see https://stackoverflow.com/questions/338166/which-direction-should-the-arrows-point-in-a-sorted-table).

I've fixed the problem verifying the last record when zoomed in, plus changed the wording on the iNat link. All are in the develop branch so I'll merge to master soon and then can setup the iNat alternative page.
